src/pudl/harvest/harvest.py

```diff
@@ -0,0 +1,1054 @@
+        together in one more input dataframes, and that each data field is in one of
+        those input dataframes. Periodic key fields (e.g. 'report_month') are matched to
+        any column of the same name with an equal or smaller period (e.g. 'report_day').
+
+        Args:
+            resources: Resources with required attributes
+                `name`, `schema`.`fields` with `name`, and `schema`.`primaryKey`.
+
+        Raises:
+            ValueError: Resource has duplicate field basenames.
+            KeyError: Missing columns needed by resource.
+        """
+        for resource in resources:
+            columns = set([field["name"] for field in resource["schema"]["fields"]])
+            basenames = [split_period(col)[0] for col in columns]
+            if len(set(basenames)) != len(basenames):
+                raise ValueError(
+                    f"Resource {resource['name']} has duplicate field basenames"
+                )
+            if resource.get("harvest"):
+                key = resource["schema"]["primaryKey"]
+                for df in self.dfs.values():
+                    df_key, df_data = find_sample(df, key=key, data=columns)
+                    if len(df_key) == len(key):
+                        columns -= set(key) | set(df_data)
+            else:
+                columns -= set(self.dfs[resource["name"]].columns)
+            if columns:
+                raise KeyError(
+                    f'Missing columns needed by resource {resource["name"]}: {columns}'
+                )
+
+    def build(
+        self,
+        resources: Iterable[dict],
+        aggregate: bool = True,
+        errors: Literal["raise", "coerce", "log"] = "raise",
+        errorfunc: Callable = None,
+    ) -> Dict[str, pd.DataFrame]:
+        """
+        Build resources.
+
+        Prunes and test resources before building (see :meth:`prune` and :meth:`test`).
+
+        For standard resources (`harvest: False`, default), the columns matching all key
+        fields and any data fields are extracted from the input dataframe of the same
+        name.
+
+        For harvested resources (`harvest: True`), the columns matching all key fields
+        and any data fields are extracted from each compatible input dataframe,
+        concatenated into a single dataframe, grouped by key fields, and aggregated
+        using the aggregate function for each data field (`schema.fields[*].aggregate`).
+        Periodic key fields (e.g. 'report_month') are matched to any column of the same
+        name with an equal or smaller period (e.g. 'report_day') and snapped to the
+        start of the desired period before aggregation.
+
+        Args:
+            resources: Resources with required attributes `name`,
+                `schema`.`fields` with `name` and `aggregate` (if `aggregate`),
+                and `schema`.`primaryKey`.
+            aggregate: Whether to aggregate harvested resources
```

Comment:
As the example in the class docstring now puts it:

> Skip aggregation to access all the rows concatenated from the input dataframes.

So you get back all the rows, not just one normalized row, for each primary key. You once asked for the ability to access the complete harvested data pre-normalization for debugging purposes. This would allow you to do that.

Reply:
If i understand correctly, the `aggregate` functionality sounds great for testing. I'd like to play with it once we have all of this hooked up to the full pudl tables.